Thomas/1584 update docs apps (#1675)
cosmos/ibc-go: docs/ibc/apps/ibcmodule.md

<!--
order: 2
-->

# Implement `IBCModule` interface and callbacks

Learn how to implement the `IBCModule` interface and all of the callbacks it requires. {synopsis}

The Cosmos SDK expects all IBC modules to implement the [`IBCModule`
interface](https://github.com/cosmos/ibc-go/tree/main/modules/core/05-port/types/module.go). This interface contains all of the callbacks IBC expects modules to implement. The include callbacks related to channel handshake, opening and closing and packet callbacks (`OnRecvPacket`, `OnAcknowledgementPacket` and `OnTimeoutPacket`).

> ```go

Author: Re: transfer, absolutely. Already cut a lot of those from the comments but some have remained while copying. will do an extra check

Re: ibc_module.go, I don't think we enforce this file structuring so I can add a suggestion but mention it's up to the dev how to structure their code?

This section will describe the callbacks that are called during channel handshake execution.

Here are the channel handshake callbacks that modules are expected to implement:

> Note that some of the code below is _pseudo code_, indicating what actions need to happen but leaving it up to the developer to implement a custom implementation. E.g. the `checkArguments(args)` function.

```go
// Called by IBC Handler on MsgOpenInit
func (im IBCModule) OnChanOpenInit(ctx sdk.Context,
    order channeltypes.Order,
    connectionHops []string,
    portID string,
    channelID string,
    channelCap *capabilitytypes.Capability,
    counterparty channeltypes.Counterparty,
    version string,
) (string,error) {
    // ... do custom initialization logic

    // Use above arguments to determine if we want to abort handshake
    // Examples: Abort if order == UNORDERED,
    // Abort if version is unsupported
    if err := checkArguments(args); err != nil {
        return "", err
    }

    if strings.TrimSpace(version) == "" {
		version = types.Version
	}

	if version != types.Version {
		return "", sdkerrors.Wrapf(types.ErrInvalidVersion, "got %s, expected %s", version, types.Version)

> 	}

Reviewer: this may not be true. Applications might generate some custom version that isn't static. Maybe replace with pseudo-code that checks the provided version for validity. ICS27 is an example of a version string that is dynamic per channel

Author: good catch, thanks!

     // OpenInit must claim the channelCapability that IBC passes into the callback

>     if err := im.keeper.ClaimCapability(ctx, chanCap, host.ChannelCapabilityPath(portID, channelID)); err != nil {

Reviewer 2: maybe link to docs about capabilities here

			return "", err
	}

    return version, nil
}

// Called by IBC Handler on MsgOpenTry
func (im IBCModule) OnChanOpenTry(
    ctx sdk.Context,
    order channeltypes.Order,
    connectionHops []string,
    portID,
    channelID string,
    channelCap *capabilitytypes.Capability,
    counterparty channeltypes.Counterparty,
    counterpartyVersion string,
) (string, error) {
    // ... do custom initialization logic

    // Use above arguments to determine if we want to abort handshake
    if err := checkArguments(args); err != nil {
        return "", err
    }

    if counterpartyVersion != types.Version {
		return "", sdkerrors.Wrapf(types.ErrInvalidVersion, "invalid counterparty version: got: %s, expected %s", counterpartyVersion, types.Version)

> 	}

Reviewer: same comment as above

    // Module may have already claimed capability in OnChanOpenInit in the case of crossing hellos
    // (ie chainA and chainB both call ChanOpenInit before one of them calls ChanOpenTry)
    // If the module can already authenticate the capability then the module already owns it so we don't need to claim
    // Otherwise, module does not have channel capability and we must claim it from IBC
    if !im.keeper.AuthenticateCapability(ctx, chanCap, host.ChannelCapabilityPath(portID, channelID)) {
        // Only claim channel capability passed back by IBC module if we do not already own it
        if err := im.keeper.scopedKeeper.ClaimCapability(ctx, chanCap, host.ChannelCapabilityPath(portID, channelID)); err != nil {
            return "", err
        }
    }

> 

Author: thanks for the headsup on this!

func (im IBCModule) OnChanOpenConfirm(
    ctx sdk.Context,
    portID,
    channelID string,
) error {
    // do custom logic

    return nil
}
```

The channel closing handshake will also invoke module callbacks that can return errors to abort the closing handshake. Closing a channel is a 2-step handshake, the initiating chain calls `ChanCloseInit` and the finalizing chain calls `ChanCloseConfirm`.

```go
// Called by IBC Handler on MsgCloseInit
func (im IBCModule) OnChanCloseInit(
    ctx sdk.Context,
    portID,
    channelID string,
) error {
    // ... do custom finalization logic

    // Use above arguments to determine if we want to abort handshake
    err := checkArguments(args)
    return err
}

// Called by IBC Handler on MsgCloseConfirm
func (im IBCModule) OnChanCloseConfirm(
    ctx sdk.Context,
    portID,
    channelID string,
) error {
    // ... do custom finalization logic

    // Use above arguments to determine if we want to abort handshake
    err := checkArguments(args)
    return err
}
```

### Channel handshake version negotiation

Application modules are expected to verify versioning used during the channel handshake procedure.

> - `onChanOpenInit` will verify that the relayer-chosen parameters

Reviewer 3: Nit: The `o` in all the `on...` callbacks that you go through here is lowercase but in the code sample above the `o` is uppercase. Maybe make it uppercase here as well?

  are valid and perform any custom `INIT` logic.
  It may return an error if the chosen parameters are invalid
  in which case the handshake is aborted.
  If the provided version string is non-empty, `onChanOpenInit` should return
  the version string if valid or an error if the provided version is invalid.
  **If the version string is empty, `onChanOpenInit` is expected to
  return a default version string representing the version(s)
  it supports.**
  If there is no default version string for the application,
  it should return an error if provided version is empty string.
- `onChanOpenTry` will verify the relayer-chosen parameters along with the
  counterparty-chosen version string and perform custom `TRY` logic.
  If the relayer-chosen parameters
  are invalid, the callback must return an error to abort the handshake.

>   If the counterparty-chosen version is not compatible with this modules

Reviewer 3: ```suggestion
  If the counterparty-chosen version is not compatible with this module's
```

  supported versions, the callback must return an error to abort the handshake.
  If the versions are compatible, the try callback must select the final version
  string and return it to core IBC.

>   `onChanOpenTry` may also perform custom initialization logic

Reviewer 3: ```suggestion
  `onChanOpenTry` may also perform custom initialization logic.
```

- `onChanOpenAck` will error if the counterparty selected version string

>   is invalid to abort the handshake. It may also perform custom ACK logic.

Reviewer 3: ```suggestion
  is invalid and abort the handshake. It may also perform custom ACK logic.
```

Versions must be strings but can implement any versioning structure. If your application plans to
have linear releases then semantic versioning is recommended. If your application plans to release
various features in between major releases then it is advised to use the same versioning scheme
as IBC. This versioning scheme specifies a version identifier and compatible feature set with
that identifier. Valid version selection includes selecting a compatible version identifier with

> a subset of features supported by your application for that version. The struct is used for this

Reviewer 3: ```suggestion
a subset of features supported by your application for that version. The struct used for this
```

> scheme can be found in `03-connection/types`.

Reviewer 3: Do you mean [this](https://github.com/cosmos/ibc-go/blob/main/modules/core/03-connection/types/version.go#L16)? Should we maybe add the link here as well?

Since the version type is a string, applications have the ability to do simple version verification
via string matching or they can use the already impelemented versioning system and pass the proto
encoded version into each handhshake call as necessary.

ICS20 currently implements basic string matching with a single supported version.

## Packet callbacks

> Just as IBC expected modules to implement callbacks for channel handshakes, IBC also expects modules to implement callbacks for handling the packet flow through a channel, as defined in the `IBCModule` interface.

Reviewer 3: ```suggestion
Just as IBC expects modules to implement callbacks for channel handshakes, IBC also expects modules to implement callbacks for handling the packet flow through a channel, as defined in the `IBCModule` interface.
```

Once a module A and module B are connected to each other, relayers can start relaying packets and acknowledgements back and forth on the channel.

![IBC packet flow diagram](https://ibcprotocol.org/_nuxt/img/packet_flow.1d89ee0.png)

Briefly, a successful packet flow works as follows:

1. module A sends a packet through the IBC module
2. the packet is received by module B
3. if module B writes an acknowledgement of the packet then module A will process the
   acknowledgement
4. if the packet is not successfully received before the timeout, then module A processes the
   packet's timeout.

### Sending packets

> Modules **do not send packets through callbacks**, since the modules initiate the action of sending packets to the IBC module, as opposed to other parts of the packet flow where msgs sent to the IBC

Reviewer 3: ```suggestion
Modules **do not send packets through callbacks**, since the modules initiate the action of sending packets to the IBC module, as opposed to other parts of the packet flow where messages sent to the IBC
```

> Note that some of the code below is _pseudo code_, indicating what actions need to happen but leaving it up to the developer to implement a custom implementation. E.g. the `EncodePacketData(customPacketData)` function.

```go
// retrieve the dynamic capability for this channel
channelCap := scopedKeeper.GetCapability(ctx, channelCapName)
// Sending custom application packet data
data := EncodePacketData(customPacketData)
packet.Data = data
// Send packet to IBC, authenticating with channelCap
IBCChannelKeeper.SendPacket(ctx, channelCap, packet)
```

::: warning
In order to prevent modules from sending packets on channels they do not own, IBC expects
modules to pass in the correct channel capability for the packet's source channel.
:::

### Receiving packets

To handle receiving packets, the module must implement the `OnRecvPacket` callback. This gets
invoked by the IBC module after the packet has been proved valid and correctly processed by the IBC
keepers. Thus, the `OnRecvPacket` callback only needs to worry about making the appropriate state
changes given the packet data without worrying about whether the packet is valid or not.

> Modules may return to the IBC handler an acknowledgement which implements the Acknowledgement interface.

Reviewer 3: ```suggestion
Modules may return to the IBC handler an acknowledgement which implements the `Acknowledgement` interface.
```

NOTE: Applications which process asynchronous acknowledgements must handle reverting state changes
when appropriate. Any state changes that occurred during the `OnRecvPacket` callback will be written
for asynchronous acknowledgements.

> Note that some of the code below is _pseudo code_, indicating what actions need to happen but leaving it up to the developer to implement a custom implementation. E.g. the `DecodePacketData(packet.Data)` function.

```go
func (im IBCModule) OnRecvPacket(
    ctx sdk.Context,
    packet channeltypes.Packet,
) ibcexported.Acknowledgement {
    // Decode the packet data
    packetData := DecodePacketData(packet.Data)

    // do application state changes based on packet data and return the acknowledgement
    // NOTE: The acknowledgement will indicate to the IBC handler if the application
    // state changes should be written via the `Success()` function. Application state
    // changes are only written if the acknowledgement is successful or the acknowledgement
    // returned is nil indicating that an asynchronous acknowledgement will occur.
    ack := processPacket(ctx, packet, packetData)

    return ack
}
```

> Reminder, the Acknowledgement interface:

Reviewer 3: ```suggestion
Reminder, the `Acknowledgement` interface:
```

```go
// Acknowledgement defines the interface used to return
// acknowledgements in the OnRecvPacket callback.
type Acknowledgement interface {
	Success() bool
	Acknowledgement() []byte
}
```

### Acknowledging packets

After a module writes an acknowledgement, a relayer can relay back the acknowledgement to the sender module. The sender module can
then process the acknowledgement using the `OnAcknowledgementPacket` callback. The contents of the

> acknowledgement is entirely upto the modules on the channel (just like the packet data); however, it

Reviewer 3: ```suggestion
acknowledgement is entirely up to the modules on the channel (just like the packet data); however, it
```
```suggestion
acknowledgement is entirely upto the modules on the channel (just like the packet data); however, it
```
